### DISCUSSION TITLE
beets-alternatives users? Streaming servers with "offline"?

### DISCUSSION THREAD

**Discussion by JOJ0** [beetbox/beets#4760] [Show and tell]
Hi, I'd like to find out if there are a lot of users of the beets-alternatives plugin lurking around in this discussion forum? I realise that it's such a useful tool and I have several use-cases for it. Basically I think it's awesome, even though I do have a few issues with it. One of them is this one: https://github.com/geigerzaehler/beets-alternatives/issues/69; Input certainly appreciated but the main thing why I'm here is actually:

My main use-case currently is to generate a complete duplicate of my library with one difference: All is mp3-vbr. This library is supposed to be used by streaming servers that for example share it to my mobile phone. That way transcoding on the fly is  not required and it helps reduce load on the serving machin. Currently I'm using a crappy piece of software called Synology DS Audio, which has all kinds of flaws but does the job. My Synology NAS is rather old and transcoding audio seems to be not an option, way too slow. Most importantly "make songs available offline" was the key feature why I started to use it. I'm thinking about replacing this whole setup with something entirely Open Source.

Maybe Airsonic is the way to go? With beetsonic plugin? Other streaming server ideas? Offline copying to my mobile is the main feature that would be required.

Tell me how you use the alternatives plugin!

**Comment by sindreruud** [comment]
Hey!

I use the beets-alternatives plugin alot :) I am a music collector but also a DJ. I let beets organize my "master" library. Afterwards I listen to my music, and I select the songs I also want in my DJ collection. I actually add them to a big playlist, run a python script to fix the paths, and then use the playlist to query the beets library to edit a custom field to library=DJ. Then I can run beets-alternatives to copy the files to the DJ library, embed art, and transcode where necessary. 
After that I am ready to use the library with my DJ software (Rekordbox). 

When it comes to streaming I have tried Plex which i quite liked, with PlexAMP. Good transcoding options also. And lately I have also tried Navidrome server with different clients (Sonixd was the most polished one IMO)

**Reply by JOJ0** [reply]
Well, awesome, thanks so much for sharing!! I'm DJing as well and have a similar setup in mind / in the works. Currently my "master" beets library is controlled by my "work" laptop (a Linux machine) and saved on the mentioned NAS.

The task of integrating the music from my "DJing computer" (a MacBook running Rekordbox or sometimes still NI Traktor) is still to be done. At least the DJing computer has a beets library on its own, so tags are quite in order I guess. I'm planning to import everything to my "master" beets library, saving it on the NAS but keep it tracked _where on the DJing machine_ the files were located exactly. I use to file my new music for DJing in folders like `spin-yy-mm` and never move them ever again, even if I tagged them (mp3tag back then, beets now) to make sure playlists and library in Rekordbox/Traktor stays in-tact.

Thanks for the ideas with `library=DJ`, I will probably do something similar, maybe `library=spin-yy-mm`, depending on which folder the track was residing on the DJing computer (let's call it "the monthly collections").

@sindreruud How are you managing the fact that DJing programs like Rekordbox usually don't like it if things are moved around? I love the ability of beets to bring folder names in order once and for all. I'm not yet sure how I could manage to use that neet beets feature on my DJing computer without breaking everything in Rekordbox.

For the streaming purposes (mainly to my phone for offline use) by now I have tested airsonic in a docker container on my Synology, which had rather poor performance, crashes and was not worth it IMO. Then I went to airsonic-advanced a more recent fork, again as a container on the NAS, which had disastrous performance, it was completely unusable.

I'm right now testing navidrome. I just found out how to import my beets smartplaylists....fingers crossed I finally find a FOSS streaming/offline-files solution that works. Thanks fo the sonixd client tip, I'll give that one a try! With my airsonic experiments I've used amperfy (iOS) which looked promising but went nuts with a crashing airsonic.

As for plex, I always read about it and researched a little but I always get mixed feelings about it. It does not feel right, it's not FOSS at all as far as I understand and I don't know if I want to put my hands in their decisions. Also I think their goals include a full fledged media center solution. I just require music. I just stream movies/shows from my syno nas via dnla, that's enough for my needs and super simple.

**Reply by arsaboo** [reply]
Plex is not FOSS, but Plexamp is the best client I have found so far for streaming. I also have Navidrome set up (along with Substreamer and Amperfy iOS clients), but I find it lacking. [Sonic analysis](https://support.plex.tv/articles/sonic-analysis-music/) on Plex takes it to a different level.

**Reply by sindreruud** [reply]
> How are you managing the fact that DJing programs like Rekordbox usually don't like it if things are moved around? I love the ability of beets to bring folder names in order once and for all. I'm not yet sure how I could manage to use that neet beets feature on my DJing computer without breaking everything in Rekordbox.

Simple, I don't move stuff! Heh, easier said then done. But I organize my music in a very simple fashion:
Album artist/Album name/track number. track title

I never have a need for reorganizing as I am happy with this setup. My primary way of interacting with my music is through the metadata (genre, BPM, date added, etc). And as such, the path/organization doesn't matter, and in return my DJ software (Rekordbox) also stays happy as things doesn't move around once I have added it. 

The only thing I'm sorely missing from my setup is a way to automatically add new tracks to the Rekordbox library once they are added to beets-alternatives. There are som Rekordbox python projects on github that interacts with the Rekordbox database/library. So I'm toying with the idea of making a plugin that would automatically update the Rekordbox library/database togheter with changes from beets-alternatives.

**Reply by JOJ0** [reply]
Ok so the good old "dont touch it" DJ-rule haha :-) Yeah same here, I always had my "monthly folders", so I dont have to rescan everything all the time. Just add a new "monthly" and never change filenames again. Just wanted to make sure there is other ways I dont know of...

Back then I didnt have beets. Sometimes I fixed my monthlies prior to rekordbox import with mp3tag, sometimes not because of lack of time. I could at least start to _always_ pretag things now, since I have beets. That would give me fixed paths as well. Historical monthlies probably will stay messy forever...

The idea with the plugin is awesome! I'll test your prototype! (even though my setup with multiple computers currently wouldnt work with it but that _might_ change in the future!)

**Comment by JOJ0** [comment]
I installed Navidrome recently and I finally found the solution I'd like to support. The project seems active, the project lead seems to be friendly and responsive. Most of all, Navidrome performs well and is focused on simplicity and performance on what I read, I love that. I have very dated hardware and it runs really really good with my mid-size library. I'm Impressed!

Nobody will convince me to support Plex as a server @arsaboo , I'm sorry :-) I would like to support a streaming server that is FOSS and would like to use analyzing features like probably "Sonic analysis" is, outside of the streaming server already. I want to use beets plus xtractor, listenbrainz, calliope, last.fm, and so on to build my personal smartplaylist system. :-)

I have a few questions about Navidrome posted over there, other than that I'm very happy with it: https://github.com/navidrome/navidrome/discussions/2326

This post _also_ was about beets-alternatives usage for providing a "pre-transcoding" to improve performance on my dated hardware. It seems Navidrome (plus the two clients : play:sub, amperfy) is quick enough to transcode on-the-fly when I download stuff to my phone, which was one of the main usecases for building such a setup. Since I have an issue with beets-alternatives/convert not supporting embedding of art that is already embedded in the originals https://github.com/geigerzaehler/beets-alternatives/issues/69 as well, I don't use beets-alternatives for this purpose anymore. It was an intermediate unpractical step that took a lot of time anyway and I now can spare it, which is pretty cool! I have other ideas for the use of beets-alternatives though and will keep reporting about it here :-)

**Reply by sindreruud** [reply]
Glad to hear you like Navidrome! One thing I have not really explored yet is playlist organization. I would really like to have a folder hierarchy of playlists. For example one folder called "Imported" with sub-playlists consisting of each month of the year. And then another folder called "Genres" which also would contain a bunch of playlist. And then maybe also a playlist called "BPM" that has playlist with ranges of BPM like 120-125 bpm.

It sounds like you also use playlists and smart playlists alot, so I am curious to hear if this suits your needs in regards to playlist managment/organization.

**Reply by JOJ0** [reply]
Sure, I'll give you some examples of what and how I do stuff with smartplaylists soonish!

**Reply by JOJ0** [reply]
BTW I'm finally starting to experiment and find a workflow for listening and adding to my DJ lib, your workflow posted in your first msg inspired me to finally get going. So you wrote:

 > I select the songs I also want in my DJ collection. I actually add them to a big playlist, run a python script to fix the paths, and then use the playlist to query the beets library to edit a custom field to library=DJ. 

Where are you listening to your music? At home on your computer (with beets access so to say) or on-the-go on your mobile accessing Navidrome? So, where do you create that big playlist? Within Navidrome/your Navidrome mobile client?

How would you bring such a playlist back into beets? Are you using this plugin? https://beets.readthedocs.io/en/stable/plugins/subsonicplaylist.html

**Reply by sindreruud** [reply]
My music resides on a TrueNAS machine at home. My music is then mounted on my Windows machine as a samba network share. In the context of managing my DJ library I prefer using my windows laptop. So to listen to the music, I have it imported in Foobar2000k which monitors the network share, and all music is automatically visible. I use a custom theme that fits my needs, and show me the relevant metadata. 

When I import music to beets I have a smartplaylist that creates a playlist where the name is the import-date. In other words, I have a list of playlists of all the days I have imported music. 
These playlists can be viewed in Foobar, and that makes it easy to listen to all the tracks I imported yesterday, for example.  

When I am ready to have listening session, I make a copy of the playlist of yesterdays import, and rename it to "2023-04-25-DJ_import". I then listen through all the tracks and remove the ones I don't need for my DJ library.

Once this new DJ playlist is ready, I run it through a Python script I made to fix the paths, and remove the remote path suffix. Next up you can reference this playlist as a list of tracks in Beets, by using: 
`$ beet ls playlist:/path/to/someplaylist.m3u`

Which also mean you can use it this way:
`beet modify playlist:/path/to/someplaylist.m3u library=DJ`

And then run the beets-alternatives to update the DJ library, transcoding and embedding artwork as needed. 

The python script also saves a playlist with absolute paths to the tracks, but pointing to beets-alternatives DJ location. So now I just drag&drop this playlist into Rekordbox, and i put it in a folder called "Imports", and the name of the playlist is the date. So now I have list of playlists in Rekordbox, corresponding to the date I imported them.

And then Rekordbox does it's thing with analyzing, finding bpm, updating smart playlists for genres, BPM ranges and other stuff. I also have some manually curated playlists in Rekordbox. 

I'm currently looking at having Beets manage all my playlist, and then find a way to automatically push these playlists to Rekordbox.  That would make it much easier to change DJ software in the future. 

Edit: Please note that I am going off my brain here because my server is turned off at the moment. I might have gotten some of the commands and work flows here slightly wrong.

**Reply by JOJ0** [reply]
Ok, got it, you basically generate an m3u file and use the playlist plugin to query. Nice! 

But it seems you do the reviewing of that playlist while on **a** computer. What I'd like achieve is to be able to do my djing selections on-the-go. I realised that subsonic supports "favoriting" of tracks. If the client supports it, those favorites are synced back to the server and could as well be seen in another subsonic client that supports it. I think I'd like to find a way to write those favorites to an m3u file and then I could do the same as you do: query using that m3u file (even if it has to be preprocessed to fix paths as you do with your python script), tag with flex attrs and so on.

Besides "favoriting" I see in some subsonci clients that they also have "starring" (1 to 5 stars), but I'm not sure yet if that is a feature for those clients only or if subsonic/navidrome would supports saving it.

**Reply by sindreruud** [reply]
> Ok, got it, you basically generate an m3u file and use the playlist plugin to query. Nice!

Yes, exactly! Be aware of UTF-8 encoding, you might have to use m3u8 instead, depending on your setup.


> But it seems you do the reviewing of that playlist while on a computer.

Yes, exclusively. My workflow is usually a 1-3 hour session where i curate these tracks. And once the "ball is rolling", I prefer to also get them into Rekordbox immediatly while they are fresh in memory, and fix the beatgrid and all that stuff.


>  What I'd like achieve is to be able to do my djing selections on-the-go.

That is on my to-do list to figure out. Hence why I was a bit curious how the playlist managment is in Navidrome. If i remember correctly it didn't quite suit my needs in the Sonixd mobile-client. However if I hear a really cool track that i suspect is not in my DJ-library I'll usually star it or add it to a temporary playlist, and then manually query it in beets and tag it with the DJ library tag. So not really sustainable for larger sessions with many tracks.

**Reply by JOJ0** [reply]
> And then Rekordbox does it's thing with analyzing, finding bpm, updating smart playlists for genres, BPM ranges and other stuff. I also have some manually curated playlists in Rekordbox.

I kind of always trying to do "smart" things with beets already but having things within Rekordbox has its strengths too! Also a reason why I still try to do most things outside of Rekordbox probably is because I remember times where it was not so easy to say "I play with _this_ DJ-system, please provide it", Traktor was huge back then and there were times where I still had to burn CDs to play them on a typical club system (Pioneer XDJ without USB). I always was lazy with my digital record collection because I played Vinyl most of the time ;-)

> I'm currently looking at having Beets manage all my playlist, and then find a way to automatically push these playlists to Rekordbox. That would make it much easier to change DJ software in the future.

Not sure if this helps but I'd like to give some examples of how I use smartplaylists in beets:

I save them in a separate yaml file that I include in ~/.config/beets/config.yaml, also it can become rather longish and you don't want that in your main beets config. Scrolling would be tedious. I save this yaml snippet file in a git repo, so I can easily reuse it on a different computer (same playlists are used for DJ.ing computer and master library computer.

Prerequisites/preprocessing required:

- Genres done with lastgenre and whatlastgenre plugins
- bpm, key done with xtractor but that could be done with a different plugin, I recently went back to using "Mixed in Key" because it's just one of the most accurate tools, even if it's not FOSS...
- mood_..., danceability and so one fields are done with github.com/adamjakab/BeetsPluginXtractor (prior to that I was using AcousticBrainz, which essentially used the same extractor tools (Essentia)
- `quality` is a flex attr coming from the inline plugin. It just tells if a track is FLAc already or a decent enough MP3 encoding. 


I often have BPM ranges first in the name. I exclude already mixed tracks (have quite some mix-cd's in the master library). I often have the same playlist twice

```
    - name: 138..150_UK-BreaksTechHardcore_Hi.m3u
      query: >
        country::"UK|GB" year:1990.. bpm:138..150 genre::'(?i)Techno|Breakbeat|Hardcore|Jungle|Electro'
        mixed::^$ bpm+ initial_key+
        quality:1

    - name: 138..150_UK-BreaksTechHardcore.m3u
      query: >
        country::"UK|GB" year:1990.. bpm:138..150 genre::(?i)Techno|Breakbeat|Hardcore|Jungle|Electro
        mixed::^$ bpm+ initial_key+
```



A list for each year stuff added:
```
    - name: Added_%time{$added,%Y}.m3u
      query: added:1990..2030
             bpm+ initial_key+
```

I sometimes even do several lists with smaller bpm-ranges (staying in +/-8 turntable pitch range):
```
    - name: 110..118_House_Hi.m3u
      query: >
        year:1987.. bpm:110..118 genre::"(?i)House|Acid Jazz|Nu Jazz|Acid"
        mixed::^$ bpm+ initial_key+
        quality:1

    - name: 118..127_House_Hi.m3u
      query: >
        year:1987.. bpm:118..127 genre::"(?i)House|Acid Jazz|Nu Jazz|Acid"
        mixed::^$ bpm+ initial_key+
        quality:1
```

Some examples of how to use mood_ stuff from xtractor:
```
    - name: Mood_Happy_Relaxed.m3u
      query: mood_happy:0.8.. mood_relaxed:0.8.. mood_aggressive:..0.4
             bpm+ initial_key+

    - name: Mood_Aggressive_Danceable.m3u
      query: mood_aggressive:0.9.. danceable:0.9..
             bpm+ initial_key+
```

A list for each year stuff added:
```
    - name: Added_%time{$added,%Y}.m3u
      query: added:1990..2030
             bpm+ initial_key+
```

Funky stuff out of my girlfriends CD collection, curated by genre tag but also by xtractors opinion ;-)
```
    - name: C_Rhythm_Blues.m3u
      query:
        - toplevel:"C CD collection" genre_rosamerica:rhy
        - toplevel:"C CD collection" genre:(Funk|Rhythm & Blues)
          bpm+ initial_key+
```

A list for each release year of Jungle records, that's a rather sophisticated one. List entries are combined with AND, sorting only has to be given _once_, because first searches are combined and then everything is sorted. That's how the smartplaylistplugin is just written. I use it as an advantage to not repeat the sorting options all over again.
```
    - name: Jungle_$original_year.m3u
      query:
        - toplevel:DrumnbassJungleDrumfunk original_year:1991..1996
          mixed::^$
        - toplevel:Bandcamp genre::"Drum And Bass|Jungle|Footwork|Juke|Hardcore|Breakbeat" bpm:150..176
          mixed::^$
        - toplevel:Bandcamp genre::"Drum And Bass|Jungle|Footwork|Juke|Hardcore|Breakbeat" bpm:75..85
          mixed::^$
        - genre::"Jungle|Drum And Bass" original_year:1991..1996
          mixed::^$
        - genre::"Jungle|Hardcore" original_year:1997..
          mixed::^$ bpm+ initial_key+

```

Regarding that last one and what you wrote in your first replay in this thread:

> Glad to hear you like Navidrome! One thing I have not really explored yet is playlist organization. I would really like to have a folder hierarchy of playlists. For example one folder called "Imported" with sub-playlists consisting of each month of the year. And then another folder called "Genres" which also would contain a bunch of playlist. And then maybe also a playlist called "BPM" that has playlist with ranges of BPM like 120-125 bpm.

Certainly such list definitions create a shitload of files and at some point having all smartplaylists in just one flat folder seems more than unpractical. I had that feeling as well already and thus tested that this acutally is also possible:
```
    - name: Jungle_by_release_year/$original_year.m3u
      query:
```

Folder is written correctly as well as the files. The problem is that the relative paths to the media files in the list seem to be hardcoded to always be relative to the playlist directory. The current plugin code doesn't recognize that we want to save the playlist in a subfolder and therefore also the media file pahts need to be fixed. 

I do think though that this wouldn't be too difficult to add as a feature to the smartplaylist plugin. By any chance, you are a Python developer? Myself, I won't get through with projects like that currently, maybe in autumn or something, we'll see. So if you are motivated to start I'm very happy to help, review and merge it :-) Joint effort?

**Reply by sindreruud** [reply]
Looks like a really solid playlist system you got going there! Thanks for sharing, great inspo for when I'm going down the same path. Now to get the clients to read them, organize it in folders and automatically fetch changes.

Are you able to use these playlist in your DJ software (Traktor?). If so, it would present a problem with beets-alternatives, as the paths in the playlist reference the main library, but the tracks reside in an alternate location. That's atleast a problem I have been struggling with. The only way around I have found is making a script that changes all the paths, which for my use case is not too hard. As my main library and alternate library have the same organization. So I just have to change the first bit of the path for each track.

 

> By any chance, you are a Python developer? Myself, I won't get through with projects like that currently, maybe in autumn or something, we'll see. So if you are motivated to start I'm very happy to help, review and merge it :-) Joint effort?

Unfortunately I am not a developer. But I have some very basic python knowledge that lets me atleast write som scripts. My day job is data analyst, so I usually use python to transform CSV files and transformation of other types of data. 

I do however want to extend my knowledge and try to make a simple plugin for Beets at some point.

**Reply by JOJ0** [reply]
> Looks like a really solid playlist system you got going there! Thanks for sharing, great inspo for when I'm going down the same path. Now to get the clients to read them, organize it in folders and automatically fetch changes.

Welcome! Yeah, I think the organizuation of beets-smartplaylists in subfolders would be very helpful. If you are bored you could write a feature request issue. I'll review and polish that request then. I got some knowledge about the smartplaylist codebase, so I might even start an implementation, if knowonenoone picks it up until I get to it...

> Are you able to use these playlist in your DJ software (Traktor?). If so, it would present a problem with beets-alternatives, as the paths in the playlist reference the main library, but the tracks reside in an alternate location. That's atleast a problem I have been struggling with. The only way around I have found is making a script that changes all the paths, which for my use case is not too hard. As my main library and alternate library have the same organization. So I just have to change the first bit of the path for each track.

Currently I'm not using those exact same smartplaylist m3u files on both machines directly. I generated them on each of them. I share a file smartplaylist.yaml in a git repo between my main lib computer and my DJ computer. Thus, I don't have a path problem. (Yeah using Rekordbox mainly recently, I even bought the Vinyl simulation system from them. Try to stick with one system and not touch Traktor anymore. Sometimes I'm thinking about switching to FOSS software mixxx, but that would mean: No easy "plug USB into Pioneer and done" anymore)

My setup is not finished yet and I might be doing it similar like you do soon, that is, using beets-alternatives. If I were to fix paths I would also just quickfix the relative paths to my needs. I recently came up with this shell/bash/zsh snippet to fix playlists, probably similar to what your Python script does (slightly different purpose but could be reused for fixing path prefixes as well):

```
function update_playlists {
    echo "Updating music-vbr playlists..."
    cp -v /remote/data/music-library/0-playlists/* /remote/data/music-vbr/0-playlists/
    
    for playlist in /remote/data/music-vbr/0-playlists/*; do
       sed -i -e 's/\.wma/.mp3/g' $playlist 
       sed -i -e 's/\.flac/.mp3/g' $playlist 
       sed -i -e 's/\.m4a/.mp3/g' $playlist 
    done
}
```
And maybe, just maybe, we should request another feature for the smartplaylist plugin to be able to generate those m3u files with different relative path prefixes.....not sure if that's a doable idea but we could try. The start of everything would be a feature request issue on here.

By any chance, you are a Python developer? Myself, I won't get through with projects like that currently, maybe in autumn or something, we'll see. So if you are motivated to start I'm very happy to help, review and merge it :-) Joint effort?

> Unfortunately I am not a developer. But I have some very basic python knowledge that lets me atleast write som scripts. My day job is data analyst, so I usually use python to transform CSV files and transformation of other types of data.

I always thought all data analyst are Python pros :-)

> I do however want to extend my knowledge and try to make a simple plugin for Beets at some point.

I would be more than happy to review such a plugin and might even be able to help and hopefully even suggest useful things :-) Yeah you should give it a try! Also regarding that library about Rekordbox interaction, please post where you found it. I'm so curious!

And furthermore about Rekordbox, what I always found supertedious is these facts (note: it's been some time since my last use of the program, new versions might have changed these behaviours, you tell me, thanks!)

- m3u files can only be imported into Rekorbox one-by-one. We can't mark multiple playlists at once and import them. Suuuuppeeeerteeediiiiioooouuuuus! :-(
- I didn't find or know of any API to interact with it or any other non-gui/scripting possiblities.

**Reply by JOJ0** [reply]
> I would be more than happy to review such a plugin and might even be able to help and hopefully even suggest useful things :-) Yeah you should give it a try! Also regarding that library about Rekordbox interaction, please post where you found it. I'm so curious!

I was curious enough to investigate a little and found this: https://pyrekordbox.readthedocs.io/en/stable/quickstart.html#rekordbox-6-database
I suppose that is what you were referring to?
 
> And furthermore about Rekordbox, what I always found supertedious is these facts (note: it's been some time since my last use of the program, new versions might have changed these behaviours, you tell me, thanks!)

I finally upgraded to latest Rekordbox (6.7.x)

> * m3u files can only be imported into Rekorbox one-by-one. We can't mark multiple playlists at once and import them. Suuuuppeeeerteeediiiiioooouuuuus! :-(

This is fixed. I can "mark" and import multiple m3u files and import at once. That is a massive relief.

> * I didn't find or know of any API to interact with it or any other non-gui/scripting possiblities.

It seems I was wrong. Thanks so much for your hints! There is pyrekordbox and it seems to either interact with the xml file database or the encrypted database already. What's your thoughts on those approaches? I think backing up the database and then working with it directly would be the more comfortable choice. I'm not sure anymore but isn't the working with the xml database always tedious because we'd have to export it from within Rekordbox first, then can edit it, then have to import it again? Sorry my Rekordbox tinkering knowledge is pretty outdated.

I just launched a `brew install SQLCipher`, let's see how that goes. Your RB is on Windows, right? Probably getting this tool installed is the tricky step. Good luck compiling!!! https://pyrekordbox.readthedocs.io/en/stable/installation.html#windows

**Reply by JOJ0** [reply]
Ah, now I realize a thing again that was always bothering me in RB: You have to delete each playlist one by one. There is no way to mark multiple playlists and hit delete for all of them. Especially when I reimport my beets generated lists, I'd like to quickly delete all of them and reimport. Do you know a way? (That is one task pyrekordbox could  do for  me hopefully)

**Reply by JOJ0** [reply]
> And maybe, just maybe, we should request another feature for the smartplaylist plugin to be able to generate those m3u files with different relative path prefixes.....not sure if that's a doable idea but we could try. The start of everything would be a feature request issue on here.

Whoops! Just wanted to let you know @sindreruud that such a "path prefix" feature already exists in the plugin: https://beets.readthedocs.io/en/latest/plugins/smartplaylist.html#configuration

**Reply by sindreruud** [reply]
> I was curious enough to investigate a little and found this: https://pyrekordbox.readthedocs.io/en/stable/quickstart.html#rekordbox-6-database
> I suppose that is what you were referring to?

Yeah, that's the one! I actually have not tried it myself yet. But reading the docs it seems to be potential for some cool integrations with Beets. For example smart-playlist integration, or adding flex-attr (or genre) into the Rekordbox MyTAG system (which is really cool!). I use the Pioneer CDJ basically as a query engine. "Give me tracks within +/- 5bpm of the current track, with genre house, mood happy that contains female.vocals". 

At the moment I add the MyTAG manually for each track, which is tedious, and it is not available to Beets and my main library. MyTAG information should live in Beets, and then get sent automatically to Rekordbox. 

> that such a "path prefix" feature already exists in the plugin:

I've tried this one. As I recall i got some issues with my workflow where i reference playlists in my query to Beets. So I would need multiple playlist folders. One "master" that beets can use for queries, and then a secondary one that references the alternative paths. I actually raised this as question over at discourse: https://discourse.beets.io/t/smart-playlists-with-backward-slashes/2250/3?u=sindreruud

And the answer I got (which I have yet to test) is under here:

> One thing you could try is the -c flag. Doing beet -c otherconfig.yaml … lets you “overlay” a second set of config options, which might work to replace your smartplaylist config with a new one.

**Reply by JOJ0** [reply]
Yeah exactly config overlays might help here. I use them often for a lot of different things.

**Reply by sindreruud** [reply]
@JOJ0 
Someone has actually started on a beets-rekordbox integration with pyrekordbox: [rkbeets](https://github.com/voigtjr/rkbeets).

Currently a bit limited, but has support for exporting from beets to rekordbox (with a --missing flag to only export new tracks). And also a import feature to get metadata from rekordbox (bpm, colour, tonality, etc) into beets. 

Looks promising and has potential for expansion.
